pkg/ucp/store/postgres/postgresclient.go

```diff
@@ -0,0 +1,422 @@
+	var timestamp *time.Time
+
+	result := store.ObjectQueryResult{}
+	for rows.Next() {
+		obj := store.Object{}
+		err := rows.Scan(&obj.ID, &obj.ETag, &obj.Data, &timestamp)
+		if err != nil {
+			return nil, err
+		}
+
+		// We could improve this by moving the filter logic to the SQL query.
+		//
+		// The problem is that the current filter logic is not well documented or tested, and
+		// we want to stay compatible with the existing implementation for now.
+		match, err := obj.MatchesFilters(query.Filters)
+		if err != nil {
+			return nil, err
+		} else if !match {
+			continue
+		}
+
+		result.Items = append(result.Items, obj)
+	}
+
+	err = rows.Err()
+	if err != nil {
+		return nil, err
+	}
+
+	if len(result.Items) < config.MaxQueryItemCount && config.MaxQueryItemCount > 0 {
+		// No more rows, so no need for pagination.
+		return &result, nil
+	}
+
+	if timestamp != nil {
+		// Will be empty if there were no rows.
+		token, err := p.createPaginationToken(*timestamp)
+		if err != nil {
+			return nil, err
+		}
+		result.PaginationToken = token
+	}
+
+	return &result, nil
+}
+
+// Save implements store.StorageClient.
+func (p *PostgresClient) Save(ctx context.Context, obj *store.Object, options ...store.SaveOptions) error {
+	if ctx == nil {
+		return &store.ErrInvalid{Message: "invalid argument. 'ctx' is required"}
+	}
+	if obj == nil {
+		return &store.ErrInvalid{Message: "invalid argument. 'obj' is required"}
+	}
+
+	parsed, err := resources.Parse(obj.ID)
+	if err != nil {
+		return &store.ErrInvalid{Message: "invalid argument. 'obj.ID' must be a valid resource id"}
+	}
+	if parsed.IsEmpty() {
+		return &store.ErrInvalid{Message: "invalid argument. 'obj.ID' must not be empty"}
+	}
+	if parsed.IsResourceCollection() || parsed.IsScopeCollection() {
+		return &store.ErrInvalid{Message: "invalid argument. 'obj.ID' must refer to a named resource, not a collection"}
+	}
+
+	normalized, err := storeutil.NormalizeResourceID(parsed)
+	if err != nil {
+		return err
+	}
+
+	config := store.NewSaveConfig(options...)
+
+	// Compute ETag for the current state of the object.
+	raw, err := json.Marshal(obj.Data)
+	if err != nil {
+		return err
+	}
+
+	obj.ETag = etag.New(raw)
+
+	// We need different SQL for the case where an etag is provided vs not provided.
+	//
+	// The key behavior difference is that if an etag is provided, we should not perform inserts, only updates.
+
+	// This is the more complex query that handles "upserts". It does not process etags.
+	sql := `
+WITH updated AS (
+	INSERT INTO resources (id, original_id, resource_type, root_scope, routing_scope, etag, resource_data)
+	VALUES ($1, $2, $3, $4, $5, $6, $7)
+	ON CONFLICT (id) 
+	DO UPDATE SET resource_data = $7
+	RETURNING id
+)
+SELECT
+CASE
+	WHEN EXISTS (SELECT 1 FROM updated) THEN 'Success'
+	WHEN EXISTS (SELECT 1 FROM resources WHERE id = $1) THEN 'ErrConcurrency'
+	ELSE 'ErrNotFound'
+END AS result;`
+
+	args := []any{
+		storeutil.NormalizePart(normalized.String()),
```

Comment:
i was wondering why we needed to normalize again since on ln323 we normalize it. Checked locally and saw that for a valid resource id the effect would be we update it with lowercase. Cool.

Reply:
Do you think it would make sense to change some of the naming of these functions? `NormalizeResourceID` and `NormalizePart` aren't doing the same thing.

Reply:
Renamed to `ConvertScopeIDToResourceID`.